src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,904 @@
+        if (r < 0)
+                return r;
+
+        ctx->nfnl = TAKE_PTR(nfnl);
+        return 0;
+}
+
+void fw_nftables_exit(FirewallContext *ctx) {
+        ctx->nfnl = sd_netlink_unref(ctx->nfnl);
+}
+
+static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+                                           const union in_addr_union *source,
+                                           unsigned int prefixlen) {
+        uint32_t mask, start, end;
+        unsigned int nplen;
+        int r;
+
+        assert(prefixlen <= 32);
+        nplen = 32 - prefixlen;
+
+        mask = (1U << nplen) - 1U;
+        mask = htobe32(~mask);
+        start = source->in.s_addr & mask;
+
+        r = sd_nfnl_nft_message_add_setelem(m, 0, &start, sizeof(start), NULL, 0);
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        end = be32toh(start) + (1U << nplen);
+        if (end < be32toh(start))
+                end = 0U;
+        end = htobe32(end);
+
+        r = sd_nfnl_nft_message_add_setelem(m, 1, &end, sizeof(end), NULL, 0);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_SET_ELEM_FLAGS, htobe32(NFT_SET_ELEM_INTERVAL_END));
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+#define NFT_MASQ_MSGS   3
+
+int fw_nftables_add_masquerade(
+                FirewallContext *ctx,
+                bool add,
+                int af,
+                const union in_addr_union *source,
+                unsigned int source_prefixlen) {
+        sd_netlink_message *transaction[NFT_MASQ_MSGS] = {};
+        size_t tsize;
+        int r;
+
+        if (!source || source_prefixlen == 0)
+                return -EINVAL;
+
+        r = sd_nfnl_message_batch_begin(ctx->nfnl, &transaction[0]);
+        if (r < 0)
+                return r;
+        tsize = 1;
+        if (add)
+                r = sd_nfnl_nft_message_new_setelems_begin(ctx->nfnl, &transaction[tsize], af, NFT_SYSTEMD_TABLE_NAME, NFT_SYSTEMD_MASQ_SET_NAME);
+        else
+                r = sd_nfnl_nft_message_del_setelems_begin(ctx->nfnl, &transaction[tsize], af, NFT_SYSTEMD_TABLE_NAME, NFT_SYSTEMD_MASQ_SET_NAME);
+
+        if (r < 0)
+                goto out_unref;
+
+        r = nft_message_add_setelem_iprange(transaction[tsize], source, source_prefixlen);
+        if (r < 0)
+                goto out_unref;
+
+        ++tsize;
+        assert(tsize < NFT_MASQ_MSGS);
+        r = sd_nfnl_message_batch_end(ctx->nfnl, &transaction[tsize]);
+        if (r < 0)
+                return r;
+        ++tsize;
+        r = nfnl_netlink_sendv(ctx->nfnl, transaction, tsize);
+
+out_unref:
+        while (tsize > 0)
+                sd_netlink_message_unref(transaction[--tsize]);
```

Comment:
in similar cases we added foobar_unref_many() wrappers, which take two params: the array and the size, and it goes through all entries and unrefs them. in this file you already have this loop twice, might be worth adding here too

Reply:
I was reluctant to add/expose a new sd-netlink api for this, but i can surely add it if thats wanted.